src/Sylius/Bundle/ApiBundle/StateProcessor/AdminUserProcessor.php

```diff
@@ -11,43 +11,46 @@
 
 declare(strict_types=1);
 
-namespace Sylius\Bundle\ApiBundle\DataPersister;
+namespace Sylius\Bundle\ApiBundle\StateProcessor;
 
-use ApiPlatform\Core\DataPersister\ContextAwareDataPersisterInterface;
+use ApiPlatform\Metadata\DeleteOperationInterface;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProcessorInterface;
 use Sylius\Bundle\ApiBundle\Exception\CannotRemoveCurrentlyLoggedInUser;
 use Sylius\Component\Core\Model\AdminUserInterface;
 use Sylius\Component\User\Model\UserInterface;
 use Sylius\Component\User\Security\PasswordUpdaterInterface;
 use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Webmozart\Assert\Assert;
 
-final class AdminUserDataPersister implements ContextAwareDataPersisterInterface
+/**
+ * @implements ProcessorInterface<AdminUserInterface>
+ */
+final class AdminUserProcessor implements ProcessorInterface
 {
     public function __construct(
-        private ContextAwareDataPersisterInterface $decoratedDataPersister,
+        private ProcessorInterface $persistProcessor,
+        private ProcessorInterface $removeProcessor,
         private TokenStorageInterface $tokenStorage,
         private PasswordUpdaterInterface $passwordUpdater,
     ) {
     }
 
-    public function supports($data, array $context = []): bool
+    public function process($data, Operation $operation, array $uriVariables = [], array $context = [])
     {
-        return $data instanceof AdminUserInterface;
-    }
+        Assert::isInstanceOf($data, AdminUserInterface::class);
 
-    public function persist($data, array $context = [])
-    {
         $this->passwordUpdater->updatePassword($data);
 
-        return $this->decoratedDataPersister->persist($data, $context);
-    }
+        if ($operation instanceof DeleteOperationInterface) {
+            if ($this->isTryingToDeleteLoggedInUser($data)) {
+                throw new CannotRemoveCurrentlyLoggedInUser();
```

Comment:
Would be cool to sync terminology.

Either
```suggestion
            if ($this->isTryingToRemoveLoggedInUser($data)) {
                throw new CannotRemoveCurrentlyLoggedInUser();
```
or
```suggestion
            if ($this->isTryingToDeleteLoggedInUser($data)) {
                throw new CannotDeleteCurrentlyLoggedInUser();
```